File: golang-backend/internal/model/commit.go
package model

import "gorm.io/gorm"

type Commit struct {
	ID           int64  `gorm:"primaryKey;autoIncrement" json:"id"`
	RepoID       string `gorm:"column:repo_id;not null;uniqueIndex:idx_repo_hash" json:"repoId"`
	CommitHash   string `gorm:"column:commit_hash;not null;uniqueIndex:idx_repo_hash" json:"commitHash"`
	AuthorName   string `gorm:"column:author_name;not null" json:"authorName"`
	AuthorEmail  string `gorm:"column:author_email;not null" json:"authorEmail"`
	CommitDate   int64  `gorm:"column:commit_date;not null;index:idx_commits_repo_date;index:idx_commits_author;index:idx_commits_date;index:idx_commits_repo_branch_date" json:"commitDate"`
	Message      string `gorm:"not null" json:"message"`
	FilesChanged int    `gorm:"column:files_changed;default:0" json:"filesChanged"`
	Insertions   int    `gorm:"default:0" json:"insertions"`
	Deletions    int    `gorm:"default:0" json:"deletions"`
	Branch       *string `json:"branch"`
	CreatedAt    int64  `gorm:"column:created_at" json:"createdAt"`

	Repository Repository `gorm:"foreignKey:RepoID;references:ID;onDelete:CASCADE" json:"-"`
}

func (Commit) TableName() string { return "commits" }

func (c *Commit) BeforeCreate(tx *gorm.DB) error {
	if c.CreatedAt == 0 {
		c.CreatedAt = NowMs()
	}
	return nil
}

File: python-backend/app/models/commit.py
"""Commit ORM model."""

from sqlalchemy import BigInteger, ForeignKey, Index, Integer, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Commit(Base):
    __tablename__ = "commits"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    repo_id: Mapped[str] = mapped_column(String, ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False)
    commit_hash: Mapped[str] = mapped_column(String, nullable=False)
    author_name: Mapped[str] = mapped_column(String, nullable=False)
    author_email: Mapped[str] = mapped_column(String, nullable=False)
    commit_date: Mapped[int] = mapped_column(BigInteger, nullable=False)
    message: Mapped[str] = mapped_column(String, nullable=False)
    files_changed: Mapped[int] = mapped_column(Integer, default=0)
    insertions: Mapped[int] = mapped_column(Integer, default=0)
    deletions: Mapped[int] = mapped_column(Integer, default=0)
    branch: Mapped[str | None] = mapped_column(String, nullable=True)
    created_at: Mapped[int] = mapped_column(BigInteger, nullable=False)

    repository = relationship("Repository", back_populates="commits")

    __table_args__ = (
        UniqueConstraint("repo_id", "commit_hash", name="commits_repo_id_commit_hash_key"),
        Index("idx_commits_repo_date", "repo_id", "commit_date"),
        Index("idx_commits_author", "author_email", "commit_date"),
        Index("idx_commits_date", "commit_date"),
        Index("idx_commits_repo_branch_date", "repo_id", "branch", "commit_date"),
    )
